frontend/src/common/queries/collections.ts

```diff
@@ -65,36 +80,93 @@ async function fetchCollection(
-): Promise<Collection | null> {
+): Promise<Map<VISIBILITY_TYPE, Collection> | null> {
   if (!id) {
     return null;
   }
 
+  const collections = new Map<VISIBILITY_TYPE, Collection>();
+
   const baseUrl = apiTemplateToUrl(API_URL + API.COLLECTION, { id });
 
   const finalUrl =
     visibility === VISIBILITY_TYPE.PRIVATE
       ? baseUrl + `?visibility=${VISIBILITY_TYPE.PRIVATE}`
       : baseUrl;
 
-  const response = await fetch(finalUrl, DEFAULT_FETCH_OPTIONS);
-  const result = await response.json();
+  let response = await fetch(finalUrl, DEFAULT_FETCH_OPTIONS);
+  let json = await response.json();
 
   if (!response.ok) {
-    throw result;
+    throw json;
   }
 
-  return result;
+  let datasetMap = new Map() as Collection["datasets"];
+  for (const dataset of json.datasets) {
+    datasetMap.set(dataset.original_id || dataset.id, dataset);
+  }
+
+  collections.set(visibility, { ...json, datasets: datasetMap });
+  if (visibility === VISIBILITY_TYPE.PRIVATE) {
+    response = await fetch(baseUrl, DEFAULT_FETCH_OPTIONS);
+    json = await response.json();
+
+    if (!response.ok) {
+      throw json;
+    }
+    datasetMap = new Map() as Collection["datasets"];
+    for (const dataset of json.datasets) {
+      datasetMap.set(dataset.original_id || dataset.id, dataset);
+    }
+
+    collections.set(VISIBILITY_TYPE.PUBLIC, {
+      ...json,
+      datasets: datasetMap,
+    });
+  }
+
+  return collections;
+}
+
+function useCollectionFetch({ id = "", visibility = VISIBILITY_TYPE.PUBLIC }) {
+  const { data: collections } = useCollections();
+  const queryCache = useQueryCache();
+  return useQuery<Map<VISIBILITY_TYPE, Collection> | null>(
+    [USE_COLLECTION, id, visibility, "FETCH"],
+    fetchCollection,
+    {
+      onSuccess: (data) => {
+        if (collections && data) {
+          queryCache.setQueryData([USE_COLLECTION, id, visibility], () => {
+            const newData = data.get(visibility) || ({} as Collection);
+            const collectionsWithID = collections.get(newData.id);
+            if (!collectionsWithID) return newData;
+            newData.is_revision = collectionsWithID.size > 1;
+
+            // check for diffs between revision and published collection
+            if (newData.is_revision && visibility === VISIBILITY_TYPE.PRIVATE) {
+              const publishedCollection =
+                data.get(VISIBILITY_TYPE.PUBLIC) || ({} as Collection);
+
+              newData.revision_diff = checkForRevisionChange(
+                newData,
+                publishedCollection
+              );
+            }
+            return newData;
+          });
+        }
+      },
+    }
+  );
 }
 
 export function useCollection({
   id = "",
   visibility = VISIBILITY_TYPE.PUBLIC,
 }) {
-  return useQuery<Collection | null>(
-    [USE_COLLECTION, id, visibility],
-    fetchCollection
-  );
+  useCollectionFetch({ id, visibility });
+  return useQuery<Collection>([USE_COLLECTION, id, visibility]);
```

Comment:
Should calls of this query invalidate the internal query? `useCollectionFetch()` is only called when the `useCollection` hook is generated

Reply:
What use cases do you have in mind that require query invalidation? Maybe you could export a function that invalidates the query so that you invalidate it for a specific use case.